src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -397,27 +370,426 @@ CollectionExpressionTypeKind.Span or
+                    var elementType = ((ArrayTypeSymbol)arrayTemp.Type).ElementType;
+
+                    // array[index] = element;
+                    expressions.Add(
+                        new BoundAssignmentOperator(
+                            expressionSyntax,
+                            _factory.ArrayAccess(arrayTemp, indexTemp),
+                            rewrittenValue,
+                            isRef: false,
+                            elementType));
+                    // index = index + 1;
+                    expressions.Add(
+                        new BoundAssignmentOperator(
+                            expressionSyntax,
+                            indexTemp,
+                            _factory.Binary(BinaryOperatorKind.Addition, indexTemp.Type, indexTemp, _factory.Literal(1)),
+                            isRef: false,
+                            indexTemp.Type));
+                });
+
+            var locals = localsBuilder.SelectAsArray(l => l.LocalSymbol);
+            localsBuilder.Free();
+
+            return new BoundSequence(
+                syntax,
+                locals,
+                sideEffects.ToImmutableAndFree(),
+                arrayTemp,
+                arrayType);
+        }
+
+        /// <summary>
+        /// Create and populate an list from a collection expression.
+        /// The collection may or may not have a known length.
+        /// </summary>
+        private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, TypeWithAnnotations elementType)
+        {
+            Debug.Assert(!_inExpressionLambda);
+            Debug.Assert(node.CollectionCreation is null);
+            Debug.Assert(node.Placeholder is null);
+
+            var elements = node.Elements;
+            var typeArguments = ImmutableArray.Create(elementType);
+            var collectionType = _factory.WellKnownType(WellKnownType.System_Collections_Generic_List_T).Construct(typeArguments);
+
+            var localsBuilder = ArrayBuilder<BoundLocal>.GetInstance();
+            var sideEffects = ArrayBuilder<BoundExpression>.GetInstance(elements.Length + 1);
+
+            int numberIncludingLastSpread;
+            bool useKnownLength = ShouldUseKnownLength(node, out numberIncludingLastSpread);
+            RewriteCollectionExpressionElementsIntoTemporaries(elements, numberIncludingLastSpread, localsBuilder, sideEffects);
+
+            bool useOptimizations = false;
+            MethodSymbol? setCount = null;
+            MethodSymbol? asSpan = null;
+
+            // Do not use optimizations in async method since the optimizations require Span<T>.
+            if (useKnownLength && elements.Length > 0 && _factory.CurrentFunction?.IsAsync == false)
+            {
+                setCount = ((MethodSymbol?)_compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_InteropServices_CollectionsMarshal__SetCount_T))?.Construct(typeArguments);
```

Comment:
In a custom BCL scenario, we'd want to check constraints. I'm okay to leave as-is as long as this was an explicit choice. It feels like the team has been more comfortable making assumptions about the well-known BCL APIs recently...